### PR TITLE
feat: Add table view screen

### PR DESCRIPTION
**Pull request by robinphillips**
# GOVAPP-38: Add table view list screen

This includes the capability to set the table view cell with a `GDSLocalisedString` and submit return that as a call-back to the place where the view model initialised from for the view controller. This is helpful because it ensures the view controller is not tightly coupled to a specific type outside of this module and therefore means it can be used flexibly in different situations and applications.

# Checklist

## Before raising your pull request:
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with ticket ID and a short description about the feature or update
      i.e. _DCMAW-222: Added ReadID SDK to iOS app_
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

- [x] Met all accessibility requirements?
    - [x] Checked dynamic type sizes are applied
    - [x] Checked VoiceOver can navigate your new code
    - [x] Checked a user can navigate only using a keyboard around your new code 

## Before merging your pull request:
- [ ] Ensure that the code coverage and SonarCloud checks have passed
- [ ] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [ ] Ran the app to ensure that no regressions have been caused by changes during code review.
- [ ] Targeted the correct branch; `develop`, `release` or `main`

### CODE REVIEW

File: Tests/GDSCommonTests/PatternTests/Mocks/MockListOptionsViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    let resultAction: (GDSLocalisedString) -> Void
+    let rightBarButtonTitle: GDSLocalisedString? = "right bar button"
+
+    let screenView: () -> Void
+    
+    func didDismiss() {
+        print("did dismiss")
+    }
+    
+
```
Comment (MarieseMikely) [inline Tests/GDSCommonTests/PatternTests/Mocks/MockListOptionsViewModel.swift:19]: There's an extra new line between these two functions - is it supposed to be like this?

**Review by MarieseMikely — APPROVED** [review]

**Comment by robinphillips** [comment]
Code coverage is lower because I have moved files for other screens into folders to organise the structure of the package better. Sonar is re-analysing the code as new.

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    let listFooter: String? = "Optional footer. Configure it on the view model in a similar way as the `body` property. The right bar button works the same way."
+    let buttonViewModel: ButtonViewModel = MockButtonViewModel(title: "Action button",
+                                                                      action: { print("button was tapped") })
+    let resultAction: (GDSLocalisedString) -> Void
+    let rightBarButtonTitle: GDSLocalisedString? = "Right bar button"
+
+    let screenView: () -> Void
+    
+    func didDismiss() {
+        print("did dismiss")
```
Comment (benmckillop) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:16]: More a question than a comment/request - do we want to have print statements in places like this? I know that its there, in a mock, just to show it works but wondering if this is the best way?

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift [hunk 3]
```diff
@@ -9,12 +9,13 @@ import UIKit
...
 ///
 /// - Returns: - The view name as a `String`
-///             isModal as a `Bool` to determine if it should be presented modally
+///             isModal as a `Bool` to determine if it should be presented modally, the default is false
 ///             the view as a `UIViewController` to push/present on the navigation stack.
 enum Screens: CaseIterable {
     case gdsInstructions
     case gdsInstructionsWithImage
     case gdsModalInfoView
+    case gdsListOptions
```
Comment (benmckillop) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:18]: 🥳

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift [hunk 4]
```diff
@@ -24,15 +25,17 @@ enum Screens: CaseIterable {
...
         case .gdsModalInfoView:
             return "Modal Info View"
+        case .gdsListOptions:
+            return "List Options"
         }
     }
     
     var isModal: Bool {
         switch self {
-        case .gdsInstructions, .gdsInstructionsWithImage:
```
Comment (benmckillop) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:32]: Do we want to remove these 2 if we are going down the route of having a default value instead? I had wanted to list them to ensure anyone adding a future one knew specifically how they were presenting it, but appreciate that maybe doesn't matter too much.

Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:32]: I figured the important thing is to explicitly capture the unusual cases (ie modal) on the basis that almost all screens won't be presented modally. It seemed like that modal case was already going to get quite long with just 3 screens there which is why I adjusted it to have a default condition.

I don't have a strong feeling either way, just concerned about it becoming a bit unwieldy as we add several more screen patterns.

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift (hunk 2 above)
Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:16]: Yeah, good thing to discuss. The `didDismiss` method is hard to test with unit tests because it is typically triggered by the `rightBarButtonItem` which I've yet to find a way to programmatically trigger the action for. Do you know a way?

If we were to add some UITests we could tap the button and then test the effect. So I left the print statement in there so users of the mock can see _something_ even if it cannot be programmatically tested.

Comment (DylanRJ) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:16]: We've triggered a bar button item programmatically in tests, checkout the `Tests/Screens/Permissions/AnalyticsPermissionViewControllerTests` file and specifically the `test_rightBarButtonItem_Contents` test which has:
```swift
_ = rightBarButton?.target?.perform(rightBarButton?.action, with: nil)
```
Not sure if this is best practice or not but seems to work, just to mention 🙂

Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:16]: Hmm, I'd tried that. Will have another look though.

Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift]: It works now so I guess it did just need the host app :) Thanks for the comments @DylanRJ and @benmckillop

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import GDSCommon
+
+struct MockListViewModel: ListOptionsViewModel {
+    let title: GDSLocalisedString = "This is the List Options screen pattern"
+    let body: String? = "This is the optional body label. If the view model property is `nil` then the label is hidden."
+    let listRows: [GDSLocalisedString] = ["Table view list item 1", "Table view list item two", "Table view list item 3", "Table view list item IV"]
+    let listFooter: String? = "Optional footer. Configure it on the view model in a similar way as the `body` property. The right bar button works the same way."
+    let buttonViewModel: ButtonViewModel = MockButtonViewModel(title: "Action button",
+                                                                      action: { print("button was tapped") })
```
Comment (DylanRJ) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:9]: just a tiny indentation one here 🙂

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+    
+    func didAppear() {
+        screenView()
+    }
+    
+    init(resultAction: ((GDSLocalisedString) -> Void)? = nil,
+         screenView: (() -> Void)? = nil,
+         dismissAction: (() -> Void)? = nil
+    ) {
```
Comment (DylanRJ) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockListViewModel.swift:27]: should this be on the line above as well 🙂

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift (hunk 4 above)
Comment (DylanRJ) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:32]: I think given the list of views we'll be demoing here it would make sense to remove and add `default:` as the vast majority won't be modal. Equally don't have a strong feeling either way but yeah might get to be quite a long list in no time 🙂

**Comment by DylanRJ** [comment]
Looking at the [Sonar Report for this PR](https://sonarcloud.io/summary/new_code?id=di-mobile-ios-common&pullRequest=7) which seems to be picking up other files in the repo (presumably because you've moved them to their own folders, which was definitely needed), there may be a quick way to get the requisite 80% coverage to merge.
Adding a test or two for the lifecycle events on [`GDSInstructionsViewController`](https://github.com/alphagov/di-mobile-ios-common/blob/feature/govapp-38-modal-screen-with-list/Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInstructions/GDSInstructionsViewController.swift) as these seem to have the [lowest coverage of any testable file](https://sonarcloud.io/component_measures?id=di-mobile-ios-common&pullRequest=7&metric=new_coverage&view=list) in the sonar report.
That may sit as a separate ticket but I think would get you over the required amount of test coverage 🙂

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockInstructionWithImageViewModel.swift [hunk 7]
```diff
@@ -1,23 +1,30 @@
 import GDSCommon
 import UIKit
 
-class MockInstructionWithImageViewModel: InstructionsWithImageViewModel {
+struct MockInstructionsWithImageViewModel: InstructionsWithImageViewModel {
```
Comment (amopatel1234) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockInstructionWithImageViewModel.swift:4]: should these changes maybe done in another PR?

**Review by amopatel1234 — COMMENTED** [review]
Overall its looks good, but there are some other changes here, its seems your adding in new .xcscheme, any idea why that is? does your branch need to be rebased?

**Comment by robinphillips** [comment]
I put the screens in folders. Since doing that, it's been a real struggle to get the test coverage over 80%. Arguably perhaps that could have been done in a separate PR but it seemed that we reached a threshold with the number of screens with this one where there needed to be some structure to the files. Moving the files around has resulted in Sonar analysis treating the moved files as new code. 

As for the new scheme, not sure what that's about. Will have a look.

**Review by DylanRJ — APPROVED** [review]
Happy to approve if the comments are resolved 🙂

**Review by DylanRJ — APPROVED** [review]
🙌
